### ISSUE TITLE
Tag 20 (T61String) not implemented - seeking to send PR

### ISSUE THREAD

**Issue by thosakwe** [wstrange/asn1lib#35]
Hi, and thanks for this package. It's just what I was looking for. I'm writing some code that can parse PEM certificate chains (I'll probably work into a PR for `asn1lib` when it's done) so I can verify some signatures for an Alexa skill.

It seems, though, that tag 20 is unimplemented, which is preventing me from decoding the signature from Amazon.

This is no problem - I can send a PR in - but I was wondering, which documents did you reference when writing this library? I'm not super familiar with BER/DER/etc., so I'm a bit at a loss as to where to start (the spec looks massive).

Thanks!

**Comment by wstrange** [comment]
Hey Tobe

Yea - this asn1 stuff is quite opaque.

I used this http://luca.ntop.org/Teaching/Appunti/asn1.html as a reference.

See section 5.16. 

I *think* (but am not sure) that the PEM format would use DER encoding. See that sample from 5.16. Note the first two bytes are the length encoding - which the library will calculate for you.

You can probably just clone one of the existing string methods (maybe the ASN1IA5String ?)

I'm taking a guess that most of the chars in that encoding are just plain old 8 bit ascii. The official encoding is https://en.wikipedia.org/wiki/ITU_T.61

**Comment by thosakwe** [comment]
Sounds good, thanks. I'll get hacking right away...

**Comment by wstrange** [comment]
My hunch:  If you clone ASN1IA5String  and change the tag to 20 - it will unblock you.  That is probably not spec compliant - but if they stick to ascii chars, it probably works.

**Comment by Ephenodrom** [comment]
> Hi, and thanks for this package. It's just what I was looking for. I'm writing some code that can parse PEM certificate chains (I'll probably work into a PR for `asn1lib` when it's done) so I can verify some signatures for an Alexa skill.
> 

@thosakwe 
Hello, please check out my util package at https://github.com/Ephenodrom/Dart-Basic-Utils
It already contains a util class for parsing PEM data. Maybe your usecase is already covered by this package.

```dart
AsymmetricKeyPair generateKeyPair({int keySize = 2048});
String formatKeyString(String key, String begin, String end,{int chunkSize = 64, String lineDelimiter = "\n"});
String generateRsaCsrPem(Map<String, String> attributes,RSAPrivateKey privateKey, RSAPublicKey publicKey);
String encodeASN1ObjectToPem(ASN1Object asn1Object, String begin, String end);
String encodeRSAPublicKeyToPem(RSAPublicKey publicKey);
String encodeRSAPrivateKeyToPem(RSAPrivateKey rsaPrivateKey);
RSAPrivateKey privateKeyFromPem(String pem);
RSAPublicKey publicKeyFromPem(String pem);
Uint8List getBytesFromPEMString(String pem);
RSAPrivateKey privateKeyFromDERBytes(Uint8List bytes);
RSAPrivateKey privateKeyFromASN1Sequence(ASN1Sequence asnSequence);
Uint8List rsaPublicKeyModulusToBytes(RSAPublicKey publicKey);
Uint8List rsaPublicKeyExponentToBytes(RSAPublicKey publicKey);
Uint8List rsaPrivateKeyToBytes(RSAPrivateKey privateKey);
ASN1Object encodeDN(Map<String, String> dn);
X509CertificateData x509CertificateFromPem(String pem);
String getSha1ThumbprintFromCertBytes(Uint8List bytes);
String getMd5ThumbprintFromCertBytes(Uint8List bytes);
```

**Comment by wstrange** [comment]
Closing as stale for now. Please feel to re-open if this becomes a priority
